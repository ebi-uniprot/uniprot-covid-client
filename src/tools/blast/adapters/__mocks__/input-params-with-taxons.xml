<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<parameters>
    <stringParameter>
        <name>program</name>
        <value>blastp</value>
    </stringParameter>
    <stringParameter>
        <name>matrix</name>
        <value>BLOSUM45</value>
    </stringParameter>
    <intParameter>
        <name>alignments</name>
        <value>100</value>
    </intParameter>
    <intParameter>
        <name>scores</name>
        <value>100</value>
    </intParameter>
    <stringParameter>
        <name>exp</name>
        <value>1e-1</value>
    </stringParameter>
    <stringParameter>
        <name>filter</name>
        <value>T</value>
    </stringParameter>
    <booleanParameter>
        <name>gapalign</name>
        <value>true</value>
    </booleanParameter>
    <stringParameter>
        <name>taxids</name>
        <value>9544,9606</value>
    </stringParameter>
    <stringParameter>
        <name>compstats</name>
        <value>F</value>
    </stringParameter>
    <intParameter>
        <name>align</name>
        <value>0</value>
    </intParameter>
    <stringParameter>
        <name>stype</name>
        <value>protein</value>
    </stringParameter>
    <stringParameter>
        <name>sequence</name>
        <value>http://wwwdev.ebi.ac.uk/Tools/services/rest/ncbiblast/result/ncbiblast-R20200618-212915-0519-97328820-np2/sequence</value>
    </stringParameter>
    <stringArrayParameter>
        <name>database</name>
        <value>
            <string>uniprotkb_refprotswissprot</string>
        </value>
    </stringArrayParameter>
</parameters>
