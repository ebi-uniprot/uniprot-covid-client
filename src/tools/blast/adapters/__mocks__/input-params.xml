<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<parameters>
    <stringParameter>
        <name>program</name>
        <value>blastp</value>
    </stringParameter>
    <stringParameter>
        <name>matrix</name>
        <value>BLOSUM62</value>
    </stringParameter>
    <intParameter>
        <name>alignments</name>
        <value>250</value>
    </intParameter>
    <intParameter>
        <name>scores</name>
        <value>250</value>
    </intParameter>
    <stringParameter>
        <name>exp</name>
        <value>10</value>
    </stringParameter>
    <stringParameter>
        <name>filter</name>
        <value>F</value>
    </stringParameter>
    <stringParameter>
        <name>compstats</name>
        <value>F</value>
    </stringParameter>
    <intParameter>
        <name>align</name>
        <value>0</value>
    </intParameter>
    <stringParameter>
        <name>stype</name>
        <value>protein</value>
    </stringParameter>
    <stringParameter>
        <name>sequence</name>
        <value>http://wwwdev.ebi.ac.uk/Tools/services/rest/ncbiblast/result/ncbiblast-R20200618-163637-0581-7413672-np2/sequence</value>
    </stringParameter>
    <stringArrayParameter>
        <name>database</name>
        <value>
            <string>uniprotkb_refprotswissprot</string>
        </value>
    </stringArrayParameter>
</parameters>
