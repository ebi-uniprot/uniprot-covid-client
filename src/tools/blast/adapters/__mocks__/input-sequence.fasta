>sp|P06787|CALM_YEAST Calmodulin OS=Saccharomyces cerevisiae (strain ATCC 204508 / S288c) OX=559292 PE=1 SV=1
MSSNLTEEQIAEFKEAFALFDKDNNGSISSSELATVMRSLGLSPSEAEVNDLMNEIDVDG
NHQIEFSEFLALMSRQLKSNDSEQELLEAFKVFDKNGDGLISAAELKHVLTSIGEKLTDA
EVDDMLREVSDGSGEINIQQFAALLSK