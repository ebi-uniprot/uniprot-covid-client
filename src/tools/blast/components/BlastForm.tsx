import React, {
  FC,
  useState,
  useEffect,
  useCallback,
  FormEvent,
  MouseEvent,
  useMemo,
  useRef,
} from 'react';
import { useDispatch } from 'react-redux';
import {
  Chip,
  SequenceSubmission,
  PageIntro,
  SpinnerIcon,
  sequenceProcessor,
} from 'franklin-sites';
import { useHistory } from 'react-router-dom';
import { sleep } from 'timing-functions';
import { v1 } from 'uuid';

import AutocompleteWrapper from '../../../uniprotkb/components/query-builder/AutocompleteWrapper';
import SequenceSearchLoader, {
  ParsedSequence,
  SequenceSearchLoaderInterface,
} from '../../components/SequenceSearchLoader';

import { addMessage } from '../../../messages/state/messagesActions';

import useReducedMotion from '../../../shared/hooks/useReducedMotion';
import useTextFileInput from '../../../shared/hooks/useTextFileInput';

import { createJob } from '../../state/toolsActions';

import { JobTypes } from '../../types/toolsJobTypes';
import { FormParameters } from '../types/blastFormParameters';
import {
  SType,
  Program,
  Sequence,
  Matrix,
  GapAlign,
  Database,
  Exp,
  Filter,
  Scores,
} from '../types/blastServerParameters';

import { LocationToPath, Location } from '../../../app/config/urls';
import defaultFormValues, {
  BlastFormValues,
  BlastFormValue,
  BlastFields,
  SelectedTaxon,
} from '../config/BlastFormData';
import uniProtKBApiUrls from '../../../uniprotkb/config/apiUrls';
import infoMappings from '../../../shared/config/InfoMappings';
import {
  MessageFormat,
  MessageLevel,
} from '../../../messages/types/messagesTypes';

import '../../styles/ToolsForm.scss';

const BLAST_LIMIT = 20;

// https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3848038/
const getAutoMatrixFor = (sequence: string): FormParameters['matrix'] => {
  if (sequence.length <= 34) {
    return 'PAM30';
  }
  if (sequence.length <= 49) {
    return 'PAM70';
  }
  if (sequence.length <= 85) {
    return 'BLOSUM80';
  }
  return 'BLOSUM62';
};

const FormSelect: FC<{
  formValue: BlastFormValue;
  updateFormValue: React.Dispatch<React.SetStateAction<BlastFormValue>>;
}> = ({ formValue, updateFormValue }) => {
  if (!formValue) {
    return null;
  }
  const label = BlastFields[formValue.fieldName as keyof typeof BlastFields];
  return (
    <section className="tools-form-section__item">
      <label htmlFor={label}>
        {label}
        <select
          value={formValue.selected as string}
          onChange={(e) =>
            updateFormValue({ ...formValue, selected: e.target.value })
          }
        >
          {formValue.values &&
            formValue.values.map((optionItem) => (
              <option
                value={String(optionItem.value)}
                key={String(optionItem.value)}
                data-testid={`${label}-${optionItem.value}`}
              >
                {optionItem.label ? optionItem.label : optionItem.value}
              </option>
            ))}
        </select>
      </label>
    </section>
  );
};

interface CustomLocationState {
  parameters?: Partial<FormParameters>;
}

const BlastForm = () => {
  // refs
  const sslRef = useRef<SequenceSearchLoaderInterface>(null);
  const fileInputRef = useRef<HTMLInputElement>(null);

  // hooks
  const dispatch = useDispatch();
  const history = useHistory();
  const reducedMotion = useReducedMotion();

  // state
  const initialFormValues = useMemo(() => {
    // NOTE: we should use a similar logic to pre-fill fields based on querystring
    const parametersFromHistoryState = (history.location
      ?.state as CustomLocationState)?.parameters;
    if (parametersFromHistoryState) {
      // if we get here, we got parameters passed with the location update to
      // use as pre-filled fields
      const formValues: Partial<BlastFormValues> = {};
      const defaultValuesEntries = Object.entries(defaultFormValues) as [
        BlastFields,
        BlastFormValue
      ][];
      // for every field of the form, get its value from the history state if
      // present, otherwise go for the default one
      for (const [key, field] of defaultValuesEntries) {
        formValues[key] = Object.freeze({
          ...field,
          selected:
            parametersFromHistoryState[
              field.fieldName as keyof FormParameters
            ] || field.selected,
        }) as Readonly<BlastFormValue>;
      }
      return Object.freeze(formValues) as Readonly<BlastFormValues>;
    }
    // otherwise, pass the default values
    return defaultFormValues;
  }, [history]);

  // used when the form submission needs to be disabled
  const [submitDisabled, setSubmitDisabled] = useState(false);
  // used when the form is about to be submitted to the server
  const [sending, setSending] = useState(false);
  // flag to see if the user manually changed the title
  const [jobNameEdited, setJobNameEdited] = useState(false);
  // store parsed sequence objects
  const [parsedSequences, setParsedSequences] = useState<ParsedSequence[]>(
    sequenceProcessor(initialFormValues[BlastFields.sequence].selected)
  );

  // actual form fields
  const [stype, setSType] = useState<BlastFormValues[BlastFields.stype]>(
    initialFormValues[BlastFields.stype]
  );
  const [program, setProgram] = useState<BlastFormValues[BlastFields.program]>(
    initialFormValues[BlastFields.program]
  );
  const [sequence, setSequence] = useState<
    BlastFormValues[BlastFields.sequence]
  >(initialFormValues[BlastFields.sequence]);
  const [database, setDatabase] = useState<
    BlastFormValues[BlastFields.database]
  >(initialFormValues[BlastFields.database]);
  const [taxIDs, setTaxIDs] = useState<BlastFormValues[BlastFields.taxons]>(
    initialFormValues[BlastFields.taxons]
  );
  const [threshold, setThreshold] = useState<
    BlastFormValues[BlastFields.threshold]
  >(initialFormValues[BlastFields.threshold]);
  const [matrix, setMatrix] = useState<BlastFormValues[BlastFields.matrix]>(
    initialFormValues[BlastFields.matrix]
  );
  const [filter, setFilter] = useState<BlastFormValues[BlastFields.filter]>(
    initialFormValues[BlastFields.filter]
  );
  const [gapped, setGapped] = useState<BlastFormValues[BlastFields.gapped]>(
    initialFormValues[BlastFields.gapped]
  );
  const [hits, setHits] = useState<BlastFormValues[BlastFields.hits]>(
    initialFormValues[BlastFields.hits]
  );

  // extra job-related fields
  const [jobName, setJobName] = useState(initialFormValues[BlastFields.name]);

  // taxon field handlers
  const updateTaxonFormValue = (path: string, id?: string) => {
    // Only proceed if a node is selected
    if (!id) {
      return;
    }

    const selected = (taxIDs.selected || []) as SelectedTaxon[];

    // If already there, don't add again
    if (selected.some((taxon: SelectedTaxon) => taxon.id === id)) {
      return;
    }

    // Truncate label: Homo sapiens (Man/Human/HUMAN) [9606] --> Homo sapiens (Man/Human/HUMAN) [9606]
    const label = path.replace(/ *\([^)]*\) */g, ' ');

    setTaxIDs({
      ...taxIDs,
      selected: [{ id, label }, ...selected],
    });
  };

  const removeTaxonFormValue = (id: string | number) => {
    const selected = (taxIDs.selected || []) as SelectedTaxon[];
    setTaxIDs({
      ...taxIDs,
      selected: selected.filter((taxon: SelectedTaxon) => taxon.id !== id),
    });
  };

  // form event handlers
  const handleReset = (event: FormEvent) => {
    event.preventDefault();

    // reset all form state to defaults
    setParsedSequences([]);

    setSType(defaultFormValues[BlastFields.stype]);
    setProgram(defaultFormValues[BlastFields.program]);
    setSequence(defaultFormValues[BlastFields.sequence]);
    setDatabase(defaultFormValues[BlastFields.database]);
    setTaxIDs(defaultFormValues[BlastFields.taxons]);
    setThreshold(defaultFormValues[BlastFields.threshold]);
    setMatrix(defaultFormValues[BlastFields.matrix]);
    setFilter(defaultFormValues[BlastFields.filter]);
    setGapped(defaultFormValues[BlastFields.gapped]);
    setHits(defaultFormValues[BlastFields.hits]);

    setJobName(defaultFormValues[BlastFields.name]);

    // imperatively reset SequenceSearchLoader... ðŸ˜·
    // eslint-disable-next-line no-unused-expressions
    sslRef.current?.reset();
  };

  // the only thing to do here would be to check the values and prevent
  // and prevent submission if there is any issue
  const submitBlastJob = (event: FormEvent | MouseEvent) => {
    event.preventDefault();

    if (!sequence.selected) {
      return;
    }

    setSubmitDisabled(true);
    setSending(true);

    // here we should just transform input values into FormParameters,
    // transformation of FormParameters into ServerParameters happens in the
    // tools middleware
    const parameters: FormParameters = {
      stype: stype.selected as SType,
      program: program.selected as Program,
      sequence: sequence.selected as Sequence,
      database: database.selected as Database,
      taxIDs: taxIDs.selected as SelectedTaxon[],
      threshold: threshold.selected as Exp,
      // remove "auto", and transform into corresponding matrix
      matrix:
        matrix.selected === 'auto'
          ? getAutoMatrixFor(sequence.selected as string)
          : (matrix.selected as Matrix),
      filter: filter.selected as Filter,
      // transform string into boolean
      gapped: (gapped.selected === 'true') as GapAlign,
      // transform string into number
      hits: parseInt(hits.selected as string, 10) as Scores,
    };

    const multipleParameters = parsedSequences.map((parsedSequence) => ({
      ...parameters,
      sequence: parsedSequence.raw as Sequence,
    }));

    // navigate to the dashboard, not immediately, to give the impression that
    // something is happening
    sleep(1000).then(() => {
      // We emit an action containing only the parameters and the type of job
      // the reducer will be in charge of generating a proper job object for
      // internal state. Dispatching after history.push so that pop-up messages (as a
      // side-effect of createJob) cannot mount immediately before navigating away.
      for (let i = 0; i < parsedSequences.length; i += 1) {
        // take extracted name by default
        let { name = '' } = parsedSequences[i];
        if (jobNameEdited) {
          // if one was submitted by user, and we only have one sequence, use it
          if (parsedSequences.length === 1) {
            name = jobName.selected as string;
          } else {
            // if we have more sequences, append a counter
            name = `${jobName.selected as string} - ${i + 1}`;
          }
        }
        dispatch(createJob(multipleParameters[i], JobTypes.BLAST, name));
      }

      history.push(LocationToPath[Location.Dashboard], {
        parameters: multipleParameters,
      });
    });
  };

  // effects
  // set the "Auto" matrix to the have the correct label depending on sequence
  useEffect(() => {
    const autoMatrix = getAutoMatrixFor(sequence.selected as string);
    setMatrix((matrix) => ({
      ...matrix,
      values: [
        { label: `Auto - ${autoMatrix}`, value: 'auto' },
        ...(matrix.values || []).filter((option) => option.value !== 'auto'),
      ],
    }));
  }, [sequence.selected]);

  const onSequenceChange = useCallback(
    (parsedSequences: ParsedSequence[]) => {
      const rawSequence = parsedSequences
        .map((parsedSequence) => parsedSequence.raw)
        .join('\n');

      if (rawSequence === sequence.selected) {
        return;
      }

      if (!jobNameEdited) {
        // if the user didn't manually change the title, autofill it
        setJobName((jobName) => {
          const potentialJobName = parsedSequences[0]?.name || '';
          if (jobName.selected === potentialJobName) {
            // avoid unecessary rerender by keeping the same object
            return jobName;
          }
          return { ...jobName, selected: potentialJobName };
        });
      }

      setParsedSequences(parsedSequences);
      setSequence((sequence) => ({ ...sequence, selected: rawSequence }));
      setSubmitDisabled(
        parsedSequences.length > BLAST_LIMIT ||
          parsedSequences.some((parsedSequence) => !parsedSequence.valid)
      );

      setSType((stype) => {
        // we want protein by default
        const selected =
          parsedSequences[0]?.likelyType === 'na' ? 'dna' : 'protein';
        if (stype.selected === selected) {
          // avoid unecessary rerender by keeping the same object
          return stype;
        }
        return { ...stype, selected };
      });
    },
    [jobNameEdited, sequence.selected]
  );

  // file handling
  useTextFileInput({
    input: fileInputRef.current,
    onFileContent: (content) => onSequenceChange(sequenceProcessor(content)),
    onError: (error) =>
      dispatch(
        addMessage({
          id: v1(),
          content: error.message,
          format: MessageFormat.POP_UP,
          level: MessageLevel.FAILURE,
        })
      ),
    dndOverlay: <span>Drop your input file anywhere on this page</span>,
  });

  const { name, links, info } = infoMappings[JobTypes.BLAST];

  return (
    <>
      <PageIntro title={name} links={links}>
        {info}
      </PageIntro>
      <form onSubmit={submitBlastJob} onReset={handleReset}>
        <fieldset>
          <section className="tools-form-section__item">
            <legend>
              Find a protein sequence to run BLAST sequence similarity search by
              UniProt ID (e.g. P05067 or A4_HUMAN or UPI0000000001).
            </legend>
            <div className="import-sequence-section">
              <SequenceSearchLoader ref={sslRef} onLoad={onSequenceChange} />
            </div>
          </section>
        </fieldset>
        <section className="text-block">
          <strong>OR</strong>
        </section>
        <fieldset>
          <section className="text-block">
            <legend>
              Enter one or more sequences ({BLAST_LIMIT} max). You may also
              <label className="tools-form-section__file-input">
                load from a text file
                <input type="file" ref={fileInputRef} />
              </label>
              .
            </legend>
            <SequenceSubmission
              placeholder="Protein or nucleotide sequence(s) in FASTA format."
              onChange={onSequenceChange}
              value={parsedSequences.map((sequence) => sequence.raw).join('\n')}
            />
          </section>
          <section className="tools-form-section">
            <FormSelect formValue={database} updateFormValue={setDatabase} />
            <section className="tools-form-section__item tools-form-section__item--taxon-select">
              <AutocompleteWrapper
                placeholder="Enter taxonomy names or tax IDs"
                url={uniProtKBApiUrls.organismSuggester}
                onSelect={updateTaxonFormValue}
                title="Restrict by taxonomy"
                clearOnSelect
              />
            </section>
            <section className="tools-form-section__item tools-form-section__item--selected-taxon">
              {((taxIDs.selected as SelectedTaxon[]) || []).map(
                ({ label, id }: SelectedTaxon) => (
                  <div key={label}>
                    <Chip
                      onRemove={() => removeTaxonFormValue(id)}
                      className="secondary"
                    >
                      {label}
                    </Chip>
                  </div>
                )
              )}
            </section>
          </section>
          <section className="tools-form-section">
            <section className="tools-form-section__item">
              <label>
                Name your BLAST job
                <input
                  name="title"
                  type="text"
                  autoComplete="off"
                  maxLength={100}
                  style={{
                    width: `${(jobName.selected as string).length + 2}ch`,
                  }}
                  placeholder="my job title"
                  value={jobName.selected as string}
                  onChange={(event) => {
                    setJobNameEdited(Boolean(event.target.value));
                    setJobName({ ...jobName, selected: event.target.value });
                  }}
                />
              </label>
            </section>
          </section>
          <details className="tools-form-advanced" open>
            <summary>
              <span>Advanced parameters</span>
            </summary>
            <section className="tools-form-section">
              {[
                [stype, setSType],
                [program, setProgram],
                [threshold, setThreshold],
                [matrix, setMatrix],
                [filter, setFilter],
                [gapped, setGapped],
                [hits, setHits],
              ].map(([stateItem, setStateItem]) => (
                <FormSelect
                  key={(stateItem as BlastFormValue).fieldName}
                  formValue={stateItem as BlastFormValue}
                  updateFormValue={
                    setStateItem as React.Dispatch<
                      React.SetStateAction<BlastFormValue>
                    >
                  }
                />
              ))}
            </section>
          </details>
          <section className="tools-form-section tools-form-section__main_actions">
            <section className="button-group tools-form-section__buttons">
              {sending && !reducedMotion && (
                <>
                  <SpinnerIcon />
                  &nbsp;
                </>
              )}
              <input className="button secondary" type="reset" />
              <button
                className="button primary"
                type="submit"
                disabled={submitDisabled}
                onClick={submitBlastJob}
              >
                {parsedSequences.length <= 1
                  ? 'Run BLAST'
                  : `BLAST ${parsedSequences.length} sequences`}
              </button>
            </section>
          </section>
        </fieldset>
      </form>
    </>
  );
};

export default BlastForm;
