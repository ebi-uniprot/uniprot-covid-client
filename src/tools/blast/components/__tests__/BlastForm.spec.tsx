import React from 'react';
import { createMemoryHistory } from 'history';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import { fireEvent, waitFor } from '@testing-library/react';
import renderWithRedux from '../../../../shared/__test-helpers__/RenderWithRedux';
import BlastForm from '../BlastForm';
import initialState from '../../../../app/state/rootInitialState';
import { mockSuggesterApi } from '../../../../uniprotkb/components/query-builder/__tests__/__mocks__/autocompleteWrapperData';

const mock = new MockAdapter(axios);
mock.onGet().reply(200, mockSuggesterApi.response);

const aaSequence = 'ABCDEFGHIJKLMNOPQRST';
const ntSequence = 'ATCGAGCGATAGCGAGGGAC';

let component;

describe('BlastForm test', () => {
  beforeEach(() => {
    const history = createMemoryHistory();
    component = renderWithRedux(<BlastForm />, {
      initialState: {
        ...initialState,
      },
      history,
    });
  });

  it('Renders the form', () => {
    const { asFragment } = component;
    expect(asFragment()).toMatchSnapshot();
  });

  it('Sets sequence type based on the sequence', () => {
    const { getByTestId } = component;
    const textArea = getByTestId('sequence-submission-input');
    fireEvent.change(textArea, { target: { value: aaSequence } });
    const stype1 = getByTestId('Sequence type-protein');
    expect(stype1.selected).toBeTruthy();
    fireEvent.change(textArea, { target: { value: ntSequence } });
    const stype2 = getByTestId('Sequence type-dna');
    expect(stype2.selected).toBeTruthy();
  });

  it.skip('Sets the program type based on the sequence', () => {});

  it.skip('Sets a name automatically', () => {});

  it('Sets the automatic matrix based on the sequence', () => {
    const { getByTestId } = component;
    const textArea = getByTestId('sequence-submission-input');
    fireEvent.change(textArea, { target: { value: aaSequence } });
    const matrix = getByTestId('Matrix-auto');
    expect(matrix.text).toEqual('Auto - PAM30');
    fireEvent.change(textArea, {
      target: { value: `${aaSequence}${aaSequence}` },
    });
    expect(matrix.text).toEqual('Auto - PAM70');
    fireEvent.change(textArea, {
      target: { value: `${aaSequence}${aaSequence}${aaSequence}` },
    });
    expect(matrix.text).toEqual('Auto - BLOSUM80');
    fireEvent.change(textArea, {
      target: {
        value: `${aaSequence}${aaSequence}${aaSequence}${aaSequence}${aaSequence}`,
      },
    });
    expect(matrix.text).toEqual('Auto - BLOSUM62');
  });

  it('Adds and removes a taxon', async () => {
    const {
      getByPlaceholderText,
      findByText,
      queryByText,
      getByTestId,
    } = component;
    const autocompleteInput = getByPlaceholderText(
      'Enter taxonomy names or tax IDs'
    );
    fireEvent.change(autocompleteInput, {
      target: { value: mockSuggesterApi.query },
    });
    const autocompleteItem = await findByText('rotavirus [1906931]');
    fireEvent.click(autocompleteItem);
    const chip = await findByText('Human rotavirus [1906931]');
    expect(chip).toBeTruthy();
    fireEvent.click(getByTestId('remove-icon'));
    await waitFor(() =>
      expect(queryByText('Human rotavirus [1906931]')).toBeFalsy()
    );
  });
});
