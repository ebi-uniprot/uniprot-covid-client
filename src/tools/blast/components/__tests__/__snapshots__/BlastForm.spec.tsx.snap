// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlastForm test Renders the form 1`] = `
<DocumentFragment>
  <div
    class="page-intro"
  >
    <h2>
      <button
        class="dropdown"
        type="button"
      >
        <svg
          fill="currentColor"
          height="20"
          viewBox="0 0 512 311"
          width="20"
        >
          <path
            d="M236 302L9 75a28 28 0 010-40L35 9C46-2 64-2 75 9l181 180L437 9c11-11 29-11 40 0l26 26c11 11 11 29 0 40L276 302a28 28 0 01-40 0z"
          />
        </svg>
         BLAST
      </button>
    </h2>
    <div
      class="intro-content false"
    >
      <p>
        The Basic Local Alignment Search Tool (BLAST) finds regions of local similarity between sequences, which can be used to infer functional and evolutionary relationships between sequences as well as help identify members of gene families.
      </p>
      <ul>
        <li>
          Start by providing a UniProt ID (UniProtKB accession/UniProtKB entry name/UniParc accession) into the input form OR a protein/nucleotide sequence.
        </li>
        <li>
          Optionally, you may change the Target database your BLAST will be run against or restrict by Taxonomy using the Taxonomy ID or name.
        </li>
        <li>
          Further advanced options can be used to change BLAST algorithm parameters.
        </li>
        <li>
          Your BLAST job will be provided a default name. You may edit this name as per your choice.
        </li>
        <li>
          Finally, click the "Run BLAST" button to submit your BLAST job.
        </li>
      </ul>
      <div
        class="intro-links"
      >
        <a
          href="/"
        >
          Help
        </a>
        <a
          href="/"
        >
          Video
        </a>
      </div>
    </div>
  </div>
  <form>
    <fieldset>
      <section
        class="tools-form-section__item"
      >
        <legend>
          Find a protein sequence to run BLAST sequence similarity search by UniProt ID (e.g. P05067 or A4_HUMAN or UPI0000000001).
        </legend>
        <div
          class="import-sequence-section"
        >
          <span
            class="search-input"
          >
            <input
              data-testid="search-input"
              placeholder="UniProt IDs"
              type="text"
              value=""
            />
            <span
              class="search-input__suffix"
              data-testid="search-input-suffix"
              role="presentation"
            >
              <svg
                fill="currentColor"
                height="14"
                viewBox="0 0 512 512"
                width="14"
              >
                <path
                  d="M505 443L405 343c-4-4-10-7-17-7h-16a208 208 0 10-36 36v16c0 7 3 13 7 17l100 100c9 9 24 9 34 0l28-28c9-10 9-25 0-34zM208 336a128 128 0 110-256 128 128 0 010 256z"
                />
              </svg>
            </span>
          </span>
        </div>
      </section>
    </fieldset>
    <section
      class="text-block"
    >
      <strong>
        OR
      </strong>
    </section>
    <fieldset>
      <section
        class="text-block"
      >
        <legend>
          Enter one or more sequences (20 max). You may also
          <label
            class="tools-form-section__file-input"
          >
            load from a text file
            <input
              type="file"
            />
          </label>
          .
        </legend>
        <textarea
          class="sequence-submission-input"
          data-testid="sequence-submission-input"
          placeholder="Protein or nucleotide sequence(s) in FASTA format."
          spellcheck="false"
        />
      </section>
      <section
        class="tools-form-section"
      >
        <section
          class="tools-form-section__item"
        >
          <label
            for="Target database"
          >
            Target database
            <select>
              <option
                data-testid="Target database-uniprotkb_refprotswissprot"
                value="uniprotkb_refprotswissprot"
              >
                UniProtKB reference proteomes + Swiss-Prot
              </option>
              <option
                data-testid="Target database-uniprotkb"
                value="uniprotkb"
              >
                UniProtKB
              </option>
              <option
                data-testid="Target database-uniprotkb_pdb"
                value="uniprotkb_pdb"
              >
                UniProtKB with 3D structure (PDB)
              </option>
              <option
                data-testid="Target database-uniprotkb_reference_proteomes"
                value="uniprotkb_reference_proteomes"
              >
                UniProtKB reference proteomes
              </option>
              <option
                data-testid="Target database-uniprotkb_swissprot"
                value="uniprotkb_swissprot"
              >
                UniProtKB Swiss-Prot
              </option>
              <option
                data-testid="Target database-UniRef100"
                value="UniRef100"
              >
                UniRef100
              </option>
              <option
                data-testid="Target database-UniRef90"
                value="UniRef90"
              >
                UniRef90
              </option>
              <option
                data-testid="Target database-UniRef50"
                value="UniRef50"
              >
                UniRef50
              </option>
              <option
                data-testid="Target database-uniparc"
                value="uniparc"
              >
                UniParc
              </option>
            </select>
          </label>
        </section>
        <section
          class="tools-form-section__item tools-form-section__item--taxon-select"
        >
          <label
            for="5"
          >
            Restrict by taxonomy
            <div
              class="autocomplete-container"
            >
              <span
                class="search-input"
              >
                <input
                  data-testid="search-input"
                  placeholder="Enter taxonomy names or tax IDs"
                  type="text"
                  value=""
                />
                <span
                  class="search-input__suffix"
                  data-testid="search-input-suffix"
                  role="presentation"
                >
                  <svg
                    fill="currentColor"
                    height="14"
                    viewBox="0 0 512 512"
                    width="14"
                  >
                    <path
                      d="M505 443L405 343c-4-4-10-7-17-7h-16a208 208 0 10-36 36v16c0 7 3 13 7 17l100 100c9 9 24 9 34 0l28-28c9-10 9-25 0-34zM208 336a128 128 0 110-256 128 128 0 010 256z"
                    />
                  </svg>
                </span>
              </span>
              <div
                class="autocomplete-menu"
              >
                <div
                  class="dropdown-menu__panel"
                />
              </div>
            </div>
          </label>
        </section>
        <section
          class="tools-form-section__item tools-form-section__item--selected-taxon"
        />
      </section>
      <section
        class="tools-form-section"
      >
        <section
          class="tools-form-section__item"
        >
          <label>
            Name your BLAST job
            <input
              autocomplete="off"
              maxlength="100"
              name="title"
              placeholder="my job title"
              style="width: 2ch;"
              type="text"
              value=""
            />
          </label>
        </section>
      </section>
      <details
        class="tools-form-advanced"
        open=""
      >
        <summary>
          <span>
            Advanced parameters
          </span>
        </summary>
        <section
          class="tools-form-section"
        >
          <section
            class="tools-form-section__item"
          >
            <label
              for="Sequence type"
            >
              Sequence type
              <select>
                <option
                  data-testid="Sequence type-protein"
                  value="protein"
                >
                  Protein
                </option>
                <option
                  data-testid="Sequence type-dna"
                  value="dna"
                >
                  DNA/RNA
                </option>
              </select>
            </label>
          </section>
          <section
            class="tools-form-section__item"
          >
            <label
              for="Program"
            >
              Program
              <select>
                <option
                  data-testid="Program-blastp"
                  value="blastp"
                >
                  blastp
                </option>
                <option
                  data-testid="Program-blastx"
                  value="blastx"
                >
                  blastx
                </option>
                <option
                  data-testid="Program-blastn"
                  value="blastn"
                >
                  blastn
                </option>
                <option
                  data-testid="Program-tblastx"
                  value="tblastx"
                >
                  tblastx
                </option>
                <option
                  data-testid="Program-tblastn"
                  value="tblastn"
                >
                  tblastn
                </option>
              </select>
            </label>
          </section>
          <section
            class="tools-form-section__item"
          >
            <label
              for="E-Threshold"
            >
              E-Threshold
              <select>
                <option
                  data-testid="E-Threshold-1e-4"
                  value="1e-4"
                >
                  0.0001
                </option>
                <option
                  data-testid="E-Threshold-1e-3"
                  value="1e-3"
                >
                  0.001
                </option>
                <option
                  data-testid="E-Threshold-1e-2"
                  value="1e-2"
                >
                  0.01
                </option>
                <option
                  data-testid="E-Threshold-1e-1"
                  value="1e-1"
                >
                  0.1
                </option>
                <option
                  data-testid="E-Threshold-1.0"
                  value="1.0"
                >
                  1
                </option>
                <option
                  data-testid="E-Threshold-10"
                  value="10"
                >
                  10
                </option>
                <option
                  data-testid="E-Threshold-100"
                  value="100"
                >
                  100
                </option>
                <option
                  data-testid="E-Threshold-1000"
                  value="1000"
                >
                  1000
                </option>
              </select>
            </label>
          </section>
          <section
            class="tools-form-section__item"
          >
            <label
              for="Matrix"
            >
              Matrix
              <select>
                <option
                  data-testid="Matrix-auto"
                  value="auto"
                >
                  Auto - PAM30
                </option>
                <option
                  data-testid="Matrix-BLOSUM45"
                  value="BLOSUM45"
                >
                  BLOSUM45
                </option>
                <option
                  data-testid="Matrix-BLOSUM62"
                  value="BLOSUM62"
                >
                  BLOSUM62
                </option>
                <option
                  data-testid="Matrix-BLOSUM80"
                  value="BLOSUM80"
                >
                  BLOSUM80
                </option>
                <option
                  data-testid="Matrix-PAM70"
                  value="PAM70"
                >
                  PAM70
                </option>
                <option
                  data-testid="Matrix-PAM30"
                  value="PAM30"
                >
                  PAM30
                </option>
              </select>
            </label>
          </section>
          <section
            class="tools-form-section__item"
          >
            <label
              for="Filter"
            >
              Filter
              <select>
                <option
                  data-testid="Filter-F"
                  value="F"
                >
                  None
                </option>
                <option
                  data-testid="Filter-T"
                  value="T"
                >
                  Filter low complexity regions
                </option>
              </select>
            </label>
          </section>
          <section
            class="tools-form-section__item"
          >
            <label
              for="Gapped"
            >
              Gapped
              <select>
                <option
                  data-testid="Gapped-true"
                  value="true"
                >
                  yes
                </option>
                <option
                  data-testid="Gapped-false"
                  value="false"
                >
                  no
                </option>
              </select>
            </label>
          </section>
          <section
            class="tools-form-section__item"
          >
            <label
              for="Hits"
            >
              Hits
              <select>
                <option
                  data-testid="Hits-50"
                  value="50"
                >
                  50
                </option>
                <option
                  data-testid="Hits-100"
                  value="100"
                >
                  100
                </option>
                <option
                  data-testid="Hits-250"
                  value="250"
                >
                  250
                </option>
                <option
                  data-testid="Hits-500"
                  value="500"
                >
                  500
                </option>
                <option
                  data-testid="Hits-750"
                  value="750"
                >
                  750
                </option>
                <option
                  data-testid="Hits-1000"
                  value="1000"
                >
                  1000
                </option>
              </select>
            </label>
          </section>
        </section>
      </details>
      <section
        class="tools-form-section tools-form-section__main_actions"
      >
        <section
          class="button-group tools-form-section__buttons"
        >
          <input
            class="button secondary"
            type="reset"
          />
          <button
            class="button primary"
            type="submit"
          >
            Run BLAST
          </button>
        </section>
      </section>
    </fieldset>
  </form>
</DocumentFragment>
`;
